src/krb5/_krb5_types.pxd

```diff
@@ -6,16 +6,29 @@ from libc.stdint cimport int32_t, uint8_t, uint32_t
 
 cdef extern from "python_krb5.h":
     """
-    // The structures are different so cannot be explicitly defined in Cython. Use inline C to set the structs elements
-    // by name.
-    void pykrb5_set_krb5_data(
+    // The structures are different so cannot be explicitly defined in Cython; use inline C.
+    int pykrb5_set_krb5_data(
         krb5_data *data,
         size_t length,
         char *value
     )
     {
+    /*
+       The caller (at least, MIT Kerberos) expects us to return the
+       data in the buffer referred to by the krb5_data struct as it's
+       passed to us -- and not, as would also be plausible, to
+       *update* the krb5_data with our *own* buffer. That, is we must
+       copy the data from value to data->data, rather than update the
+       data->data pointer. If we do the latter, the caller simply
+       ignores the new buffer we return and proceeds to use its own
+       buffer, which we never updated.
+    */
+        if (data->length < length)
```

Comment:
Why not put this check in the Python code in `prompt_callback`. This way the C code stays a simple stub and the Cython code is the one doing the validation/raising the exception.

Reply:
I have a mild preference for doing it this way, since the validation is intimately tied to the nature of the `krb5_data` update semantics implemented right below it; it seems more natural and somewhat safer to do it here. But I defer to whatever you prefer. It does mean the C routine could remain `void` and not have to return anything.

Reply:
I suppose it actually needs to be done in the C side as you can't access the length field of the struct through Cython. Ignore this suggestion :)